File: src/1930_unique_length_3_palindromic_subsequences.js
/**
 * @param {string} s
 * @return {number}
 */
var countPalindromicSubsequence = function (s) {
    let count = 0;
    let chars = new Set(s);
    for (const char of chars) {
        let first = s.indexOf(char),
            last = s.lastIndexOf(char);
        count += new Set(s.slice(first + 1, last)).size;
    }
    return count;
};

module.exports = { countPalindromicSubsequence };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(countPalindromicSubsequence("aabca"), 3);
    assert.deepStrictEqual(countPalindromicSubsequence("adc"), 0);
    assert.deepStrictEqual(countPalindromicSubsequence("bbcbaba"), 4);
}

File: src/1930_unique_length_3_palindromic_subsequences.py

# LeetCode 1930: Unique Length 3 Palindromic Subsequences
# Python version

def charIndex(ch: str) -> int:
    return ord(ch) - ord("a")

def countPalindromicSubsequence(s: str) -> int:
    count = 0
    chars = set(s)
    for char in chars:
        first,last = s.find(char),s.rfind(char)
        middle = [False] * 26
        i = first + 1
        while i < last:
            middle[charIndex(s[i])] = True
            i += 1
        for seen in middle:
            if seen:
                count += 1
    return count


def main():
    assert countPalindromicSubsequence('aabca') == 3
    assert countPalindromicSubsequence('adc') == 0
    assert countPalindromicSubsequence('bbcbaba') == 4

if __name__ == "__main__":
    main()
